### PR TITLE
Turn off some jobs in PR CI that only differ on Python version

### PR DESCRIPTION
**Pull request by yf225**
... to make CI less flaky. https://github.com/pytorch/pytorch/issues/16459

### CODE REVIEW

**Comment by ezyang** [comment]
OK, so the hypothesis is that if there are Python 3 errors, we'll catch them in the CUDA build?

**Review by ezyang — APPROVED** [review]
This seems like a reasonable start, but I feel like we can turn off more configs. The easiest way to get information about this is to do a little bit of data analysis about how correlated our configs are, group them into groups ,and then pick one representative from each group.

Barring that, my unscientific suggestion is that we test:
* One mobile build (Caffe2)
* Two Linux CPU builds (PyTorch, Caffe2)
* Two Linux GPU builds (PyTorch, Caffe2+ONNX)
* Two OS X builds (PyTorch, Caffe2)
* Two Windows GPU builds (PyTorch, Caffe2)

@pjh5 My suggestion is that we don't include binary builds in this set. What do you think? When we merge PyTorch and Caffe2 tests, we can further reduce this configuration by four.

We should also add comments in this yml file explaining to people how to enable all tests when they need to on their PR. Ideally no yml editing would be necessary to trigger this behavior, but CircleCI might not give us enough rope to do this using the `filters` mechanism.

**Comment by yf225** [comment]
@ezyang I made some changes to the PR and here is a list of jobs that are always run:
```
# Pytorch linux builds
- pytorch_linux_trusty_py2_7
- pytorch_linux_trusty_pynightly
- pytorch_linux_xenial_cuda8_cudnn7_py3

# Pytorch MacOS builds
- pytorch_macos_10_13_py3

# Caffe2 Linux builds
- caffe2_cmake_cuda9_0_cudnn7_ubuntu16_04
- caffe2_py2_mkl_ubuntu16_04
- caffe2_py2_cuda9_0_cudnn7_centos7
- caffe2_onnx_py2_gcc5_ubuntu16_04

# Caffe2 MacOS builds
- caffe2_py2_system_macos10_13

# Caffe2 mobile builds
- caffe2_py2_android_ubuntu16_04
- caffe2_py2_ios_macos10_13

# Binary builds
- binary_linux_manywheel_2.7mu_cpu_build
- binary_linux_manywheel_2.7mu_cpu_test:
    requires:
        - binary_linux_manywheel_2.7mu_cpu_build
- binary_linux_conda_3.7_cu100_build
- binary_linux_conda_3.7_cu100_test:
    requires:
        - binary_linux_conda_3.7_cu100_build
- binary_macos_conda_3.5_cpu_build
```

A few of the Caffe2 Linux builds differ on some interesting dimensions (cmake / mkl / centos / onnx), so I keep them in the PRs in case we can catch something interesting there. We can look at this list again when we are able to further reduce the number of variants.

I also added `PR_CI_COVERAGE_SETTING` that takes `run_all_jobs` or `run_high_signal_jobs_only` as values for people to toggle between full CI matrix / high-signal jobs, and a CI job `please_revert_change_to_PR_CI_COVERAGE_SETTING` to throw an error if people have `PR_CI_COVERAGE_SETTING` set to `run_all_jobs` in their PR (in case they forgot to set it back when they are done with testing the full CI matrix).

**Comment by pjh5** [comment]
We do occasionally have failures in the binary builds only on conda/pip python3/2 . Moving them to on-merge instead of on-PR is probably fine, especially after T39509675

File: .circleci/config.yml [hunk 1]
```diff
@@ -3006,69 +3022,38 @@ workflows:
...
             - pytorch_macos_10_13_py3_build
-      - pytorch_macos_10_13_cuda9_2_cudnn7_py3_build
 
-      # Caffe2 builds
+      # Caffe2 Linux builds
       - caffe2_cmake_cuda9_0_cudnn7_ubuntu16_04_build
       - caffe2_cmake_cuda9_0_cudnn7_ubuntu16_04_test:
           requires:
             - caffe2_cmake_cuda9_0_cudnn7_ubuntu16_04_build
-      - caffe2_py2_cuda9_0_cudnn7_ubuntu16_04_build
```
Comment (pjh5) [inline .circleci/config.yml]: This is actually the only TensorRT build https://github.com/pytorch/pytorch/blob/master/.jenkins/caffe2/build.sh#L137

Comment (pjh5) [reply .circleci/config.yml]: A little related to this, we should eliminate all string checks on the entire config name like this, and include all needed config info the job name. cc @kostmo

**Pull request closed without merging**